Issue #2204: Add 'baseName' option for TranslationCheck

MEZk opened the pull request:
Issue https://github.com/checkstyle/checkstyle/issues/2204 .

@romani 
1) Since TranslationCheck extends AbstractFileSetCheck, it also inherits 'setFileExtensions' method which allows user to specify file extensions. As was discussed at Issue https://github.com/checkstyle/checkstyle/issues/2204 we should validate files with '.properties' extension. Should we remove the description of the option from xdoc and filter files by extension. for example, in 'processFiltered' method with help of  CommonUtils#matchesFileExtension ?

2) I think it is worth changing the issue title.

romani commented:
windows CI is not happy, please fix.

romani commented:
please rebase your PR, to resolve Travis failure.

romani commented:
http://checkstyle.sourceforge.net/config_misc.html#Translation

> . Should we remove the description of the option from xdoc and filter files by extension.

nothing should be removed, users are not aware of hierarchy that we have in Checks.

MEZk commented:
@romani 
>nothing should be removed, users are not aware of hierarchy that we have in Checks.

How should we forbid users to use 'fileExtensions' option?

romani commented:
> How should we forbid users to use 'fileExtensions' option? 

why we should do this ?

romani commented:
if you do PR rebase, PR validation should pass, now it should support both , short and long reference of issue.

Travis is failing with broken UTs too.

MEZk commented:
@romani 
Fixed.
Added possibility to validate files with different extensions.

There are not many examples of how developers use resource bundles for translation purposes on GitHub. I found not many projects which use bundle structres as the following:

baseName.properties
baseName_de_DE.properties
baseName_en_EN.properties

The most common extension is *.properties. I did not find some other extensions, but developers do use country, variant in their projects.

Finally, I think, that if there is a possibility to implement your own  Control for ResourceBundle and to organize collecting of resource files to a bundle the way you need and with extensions you need, we should validate different extensions.

romani commented:
> The most common extension is *.properties. I did not find some other extensions, but developers do use country, variant in their projects.

It is completely ok for us to force user to use extension "properties". We need from github examples of basenames and make sure we could cover their files by our Check.
we should more

MEZk commented:
@romani 

1). [SWB](https://github.com/InfotecMX/swb/tree/685dc532e176aec16e37e351064baeeff096088a/SWB4/swb/applets/WorkFlowAdmin/src/applets/workflowadmin)

Base names: JActivityFlowPanel, jTableActivitiesModel, GraphView, ZULResource,  	jTableResourceTypeModel, jTableUserModel, EditWorkflow, AuthorActivity,  	AddEditActivitySecuence, ActivityEdition, Banner, Comment, CommentSwf,  FrameContent 	JSPResourceType.

Base name regexp for the project will be too broad and complex. The base names have the same names as java classes in the project. Proposal: add new option to specify paths for validation.

2). [ztsr-j2ee-svn-to-git](https://github.com/BGCX262/ztsr-j2ee-svn-to-git/tree/df45fa389cea5162da589b6aff0da9d50167ca07/trunk/ELMS/ELMS-war/src/java/uj/edu/elms)

Base name:  messages.
Configuration:

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name = "Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>
    <module name="TreeWalker">
         <module name="Translation">
             <property name="baseName" value="^messages$"/>
             <property name="requiredTranslations" value="pl"/>
         </module>
    </module>
</module>
```

3). [PortableSigner2](https://github.com/pflaeging/PortableSigner2/tree/023c915a7ce8aa20d1b2c2c1bc3c935842fb2872/src/net/pflaeging/PortableSigner)

Base names: Signatureblock, ToolTips, i18n.
Configuration:

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name = "Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>
    <module name="TreeWalker">
         <module name="Translation">
             <property name="baseName" value="^(Signatureblock|ToolTips|i18n)$"/>
             <property name="requiredTranslations" value="cs, de, en, es, it, pl"/>
         </module>
    </module>
</module>
```

4). [secureOWS](https://github.com/camptocamp/secureOWS/tree/314dc306bdfd0726cd750359f0dcf746d9c87205/owsproxyclient/netbeans/owsproxyclient/src/owsproxyclient)

Base name: translations.
Configuration:

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name = "Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>
    <module name="TreeWalker">
         <module name="Translation">
             <property name="baseName" value="^translations$"/>
             <property name="requiredTranslations" value="de, fr, it"/>
         </module>
    </module>
</module>
```

5). [Infor](https://github.com/daranedag/Infor/tree/af4d4b92c2899aa292b3013db8f8b0a42f0b1e52/forestsimulator75/treegross_scr/standsimulation)

Base names: EditorPanel, MyMenubar,TgDesign, TgGrafik, TgGrafikMenu, TgJFrame, TgPPMapMenu, TgStandInfo, TgUserDialog.
Configuration:

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name = "Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>
    <module name="TreeWalker">
         <module name="Translation">
             <property name="baseName" value="^(EditorPanel|MyMenubar|Tg[A-Z][a-zA-z]+)$"/>
             <property name="requiredTranslations" value="de, en, es"/>
         </module>
    </module>
</module>
```

6). [zonaclic](https://github.com/projectestac/zonaclic/tree/40bc27936d5fa34f1702d514a1281a4a9efe870c/app/clicDB/src/java/edu/xtec/resources/messages)

Base names: activityMessages, countMessages, listMessages, rssMessages.
Configuration:

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name = "Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>
    <module name="TreeWalker">
         <module name="Translation">
             <property name="baseName" value="^.+Messages$"
             <property name="requiredTranslations" value="ca, en, es"/>
         </module>
    </module>
</module>
```

7). [OpenLanguageTools](https://github.com/oncever/OpenLanguageTools/tree/8f0dd5451f2c32f8256839998bdd3431d78c9854/src/transeditor/i18n/cs_CZ/org/jvnet/olt/editor/translation)

Base names: AlignmentMain, CustomKeyboard, SourceContextFrame, OpenFileFilters.
Base name regexp for the project will be too broad. The base names have the same names as java classes in the project. Proposal: add new option to specify paths for validation.

8). [wemooc](https://github.com/manuelsanchezortiz/wemooc/tree/b380251e74dc8e6ad3b6a42b0416284108927b42/liferaylms-portlet/docroot/WEB-INF/src/content/usercompetence)

Base name: Language.
Configuration:

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name = "Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>
    <module name="TreeWalker">
         <module name="Translation">
             <property name="baseName" value="^Language$"
             <property name="requiredTranslations" value="ca, de, en, es, eu, pt, gl"/>
         </module>
    </module>
</module>
```

9). [mysakai](https://github.com/huhongjun/mysakai/tree/a0d09dde88199fd62d63ee39ac2dddc7da1d4dc6/sakai/i18n/sakai/polls/tool/src/bundle/org/sakaiproject/poll/bundle)

Base name: Messages.
Configuration:

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name = "Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>
    <module name="TreeWalker">
         <module name="Translation">
             <property name="baseName" value="^Messages$"
             <property name="requiredTranslations" value="ca, es, nl, zh"/>
         </module>
    </module>
</module>
```

romani commented:
thanks a lot for investigation. It shows that users choose any simple way to name a file, without any naming notation. They have no incentive to make any notations. But as they try to take control of this , for them introduction of naming convention will cost nothing, just a quick update on files to put any distinctive suffix to basename. 

>  Proposal: add new option to specify paths for validation.

I do not fill that is required. Lets try to force user to make localization property files to have distinctive name suffix/prefix.

romani commented:
Code review:

1)
>  checkConfig.addAttribute("baseName", "^.*(Titles|Labels).*$");

All new UT method have regexp as "^.*xxxx.*$", it is not clear why we need second ".*" ?
Please remove in all cases. User need clearly understand what is going to be matching to "basename".
If you want to keep this, please add one more UT with name like "MyLabelsI18.properties" to reuse second ".*" and show how it is matching.

2)
Please add to javdoc and xdoc example of usage fileExtension property, and put it in list of Check properties (user is not aware of how class hierarchy is done for Checks, and he should not be aware , we could change this at any time). The same for xdoc.

3)
> 
```
private static final Pattern LANGUAGE_COUNTRY_VARIANT =
        CommonUtils.createPattern("^.*\\_[a-z]{2}\\_[A-Z]{2}\\_[A-Za-z]+\\..+$");

removePattern = "\\_[a-z]{2}\\_[A-Z]{2}\\_[A-Za-z]+\\..+$";
```
copy-paste of complicated regexps, not good.
make it as : 
```
final static regexp_contry_variant = ""^.*\\_[a-z]{2}\\_[A-Z]{2}\\_[A-Za-z]+\\..+$""
please use removePattern = regexp_contry_variant.replace("^.*","")
```

MEZk commented:
1) Removed. Added new UT.
2) Done.
3) Done, but IDEA's violation appeared.


**ID**
DynamicRegexReplaceableByCompiledPattern

**Description**
Reports calls to the regular expression methods of java.lang.String using constants arguments. Such calls may be profitably replaced with a private static final Pattern field so that the regular expression does not have to be compiled each time it is used.

Should I suppress the violation? 
Another way out is to pre-compile 3 remove patterns and use the following ugly code just to satisfy IDEA's inspection that would be incorrect:

````java
Pattern.compile(removePattern).matcher(fileName).replaceAll("");
````

I also decided to remove dependency on Guava and replace StringUtils#removePattern with native java approach.

4) Done.
5) Done. 

5.1) I had to rename 'general.fileNotFound', 'general.exception' to 'tranlsation.fileNotFound', 'tranlsation.exceptionGeneral' and move them into 'com/puppycrawl/tools/checkstyle/checks/messages.properties' resource bundle. This kind of hack allowed me to avoid NPE in [CheckUtil#getCheckMessage](https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java#L193), because 'Properties#getProperty' method returns null if the property is not found and it leads to NPE in 'getCheckMessage'.

5.2) testAllCheckSections fails the build with the following message:

```config_misc.xml section 'Translation' should have the expected error keys expected:<translation.[exceptionGeneral(..)```

but, translation.exceptionGeneral is described in "Error messages" section.

romani commented:
Intelij Idea is right,  such a simple launch of String method make a runtime compilation of RegExp - http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/8u40-b25/java/lang/String.java#2218 .That is performance problem.
Please use `regexp.substring("^.*".length, regexp.length)`  .

>  I had to rename 'general.fileNotFound', 'general.exception' to 'tranlsation.fileNotFound', 'tranlsation.exception' and move them into 'com/puppycrawl/tools/checkstyle/checks/messages.properties' resource bundle

No, such update is not OK, please revert your changes for this point. Please create another issue on this. Our update for a Check is too complicated to mix it with that update, all should be done separately.

MEZk commented:
@romani 
>Please use regexp.substring("^.*".length, regexp.length) .

Done.

>No, such update is not OK, please revert your changes for this point. Please create another issue on this. 

Done. See https://github.com/checkstyle/checkstyle/issues/3015

romani commented:
6)
> 
```
protected void processFiltered(File file, List<String> lines) {
        filesToProcess.add(file);
```

please put a comment above `filesToProcess.add(file);` that we just collecting files for processing at `finishProcessing`

7)
>
```
arrangeFilesByExtensions(filesToProcess);
arrangeFilesIntoBundles(filesWithSimilarExtensions, baseNamePattern);
```

please explain a reason of pre-arrangement by Extension. if no reason please use only `arrangeFilesIntoBundles`.


8)
>
```
            checkExistenceOfTranslations(bundlesMap);
            checkPropertyFileSets(bundlesMap);
```

second method should be `checkTranslations`

9)
> `final String baseName = removePath(fullBundleName);`

rename `removePath` to `getFilename`. 

10)
>
```
        final String missingFileName = getMissingFileName(filesInResourceBundle, baseName, null);
        if (missingFileName != null) {
```

is it a Guava Option candidate ?

11)
>
```
            final String missingTranslationFileName = String.format(Locale.ROOT,
                DEFAULT_TRANSLATION_FILE_NAME_FORMATTER, fullBundleName, extension);
```

please move to special method `getTranslationFileName(fullBundleName, extension)`. rename `missingTranslationFileName` to `filename`. 
The same code in methods below, please apply this note to it too.

MEZk commented:
7)
>please explain a reason of pre-arrangement by Extension. if no reason please use only arrangeFilesIntoBundles.

These two resource bundles should be considered as different resource bundles:
```
com/somepackage/messages.translation
com/somepackage/messages_de.translation

com/somepackage/messages.properties
com/somepackage/messages_de.properties
```
They have simmilar base names. but different extensions and for Java they are two different resource bundles. If we do not pre-arrange the files by extension we will treat the files as one resource bundle.

romani commented:
It is not a reason, it should be part of arrange by basename algorithm

MEZk commented:
@romani 
We cannot distinguish two resource bundles from the example, because we use path + base name as a key for arrangement Map, either we cannot use path + base name + extension as a key, because default translation file may not be present at all.

romani commented:
all I want to do is to, make method looks like:

```
    public void finishProcessing() {
        super.finishProcessing();
        final SetMultimap<String, File> bundlesMap =
                arrangeFilesIntoBundles(arrangedFiles, baseNamePattern);
        checkExistenceOfTranslations(bundlesMap);
        checkPropertyFileSets(bundlesMap);
    }
```

and after this update it looks like:
```
    public void finishProcessing() {
        super.finishProcessing();
        final SetMultimap<String, File> bundlesMap =
                arrangeFilesIntoBundles(arrangedFiles, baseNamePattern);
        for (String fullBundleName : bundlesMap.keySet()) {
            final Set<File> filesInBundle = translations.get(fullBundleName);
            checkExistenceOfDefaultTranslation(filesInBundle, fullBundleName);
            checkExistenceOfRequiredTranslations(filesInBundle, fullBundleName);
            checkPropertyFileSets(filesInBundle);
        }
    }
```

Such method will easily describe main algorithm of Check:
1) grouping files for processing (arrangement), for each group do "2), 3), 4)"
2) checkExistenceOfDefaultTranslation
3) checkExistenceOfRequiredTranslations
4) checkPropertyFileSets

MEZk commented:
@romani 
6) Done.
7) Done.
8) Done.
9) Done.
10) Done.
11) Done.

romani commented:
merged as FF
